src/ucp/wireup/wireup_cm.c

```diff
@@ -229,37 +227,226 @@ static void ucp_cm_client_connect_cb(uct_ep_h ep, void *arg,
+        goto err_free_sa_data;
+    }
+
+    progress_arg->ucp_ep    = ucp_ep;
+    progress_arg->uct_cm_ep = uct_cm_ep;
+    memcpy(progress_arg->dev_addr, remote_data->dev_addr,
+           remote_data->dev_addr_length);
+    memcpy(progress_arg->sa_data, remote_data->conn_priv_data,
+           remote_data->conn_priv_data_length);
+
+    uct_worker_progress_register_safe(worker->uct,
+                                      ucp_cm_client_connect_progress,
+                                      progress_arg, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                      &prog_id);
+    ucp_worker_signal_internal(ucp_ep->worker);
+    return;
+
+err_free_sa_data:
+    ucs_free(progress_arg->sa_data);
+err_free_arg:
+    ucs_free(progress_arg);
+err_out:
+    ucp_ep->flags &= ~UCP_EP_FLAG_LOCAL_CONNECTED;
+    ucp_worker_set_ep_failed(worker, ucp_ep, uct_cm_ep,
+                             ucp_ep_get_cm_lane(ucp_ep), status);
 }
 
-static void ucp_cm_disconnect_cb(uct_ep_h ep, void *arg)
+/*
+ * Internal flush completion callback which is a part of close protocol,
+ * this flush was initiated by remote peer in disconnect callback on CM lane.
+ */
+static void ucp_ep_cm_disconnect_flushed_cb(ucp_request_t *req)
 {
-    ucs_error("UCP close protocol is not completely implemented");
+    ucp_ep_h ucp_ep = req->send.ep;
+
+    UCS_ASYNC_BLOCK(&ucp_ep->worker->async);
+    ucp_ep_cm_disconnect_cm_lane(ucp_ep);
+    ucp_request_complete_send(req, UCS_OK);
+    UCS_ASYNC_UNBLOCK(&ucp_ep->worker->async);
+}
+
+static unsigned ucp_ep_cm_remote_disconnect_progress(void *arg)
+{
+    ucp_ep_h ucp_ep = arg;
+    void *req;
+
+    ucs_trace("ep %p: flags %xu cm_remote_disconnect_progress", ucp_ep,
+              ucp_ep->flags);
+
+    ucs_assert(ucp_ep_get_cm_uct_ep(ucp_ep) != NULL);
+
+    ucs_assert(ucp_ep->flags & UCP_EP_FLAG_LOCAL_CONNECTED);
+    if (ucs_test_all_flags(ucp_ep->flags, UCP_EP_FLAG_CLOSED |
+                                          UCP_EP_FLAG_CLOSE_REQ_VALID)) {
+        ucp_request_complete_send(ucp_ep_ext_gen(ucp_ep)->close_req.req, UCS_OK);
+        return 1;
+    }
+
+    if (ucp_ep->flags & UCP_EP_FLAG_CLOSED) {
+        return 1;
+    }
+
+    /*
+     * TODO: set the ucp_ep to error state to prevent user from sending more
+     *       ops.
+     */
+    ucs_assert(ucp_ep->flags & UCP_EP_FLAG_FLUSH_STATE_VALID);
+    ucs_assert(!(ucp_ep->flags & UCP_EP_FLAG_CLOSED));
+    req = ucp_ep_flush_internal(ucp_ep, UCT_FLUSH_FLAG_LOCAL, NULL, 0, NULL,
+                                ucp_ep_cm_disconnect_flushed_cb,
+                                "cm_disconnected_cb");
+    if (req == NULL) {
+        /* flush is successfully completed in place, notify remote peer
+         * that we are disconnected, the EP will be destroyed from API call */
+        ucp_ep_cm_disconnect_cm_lane(ucp_ep);
+    } else if (UCS_PTR_IS_PTR(req)) {
+        /* flush is in progress, wait its completion */
+        ucp_request_release(req);
```

Comment:
ucp_request_free() ?

Reply:
no, the request will freed on completion

Reply:
@evgeny-leksikov https://github.com/openucx/ucx/blob/master/src/ucp/api/ucp_compat.h#L39

Reply:
But I don't need to disable callback, where is a difference
```
UCS_PROFILE_FUNC_VOID(ucp_request_release, (request), void *request)
{
    /* mark request as released */
    ucp_request_release_common(request, 0, "release");
}

UCS_PROFILE_FUNC_VOID(ucp_request_free, (request), void *request)
{
    /* mark request as released and disable the callback */
    ucp_request_release_common(request, UCP_REQUEST_FLAG_CALLBACK, "free");
}
```

Reply:
does it make sense to release a request but not its callback?

Reply:
why not? according to doc and implementation the request will be released after its completion callback.

Reply:
can we put (=release) the request from ucp_ep_cm_disconnect_flushed_cb?

Reply:
we can but why?
 - `ucp_request_complete` anyway does all this after invoke CB if request is released.
 - IMO use "put" from completion CB is not safety in common case because `ucp_request_complete` does access to the request. It would work for flush just because it calls CB directly from `ucp_flush_check_completion`.
 - Async is already blocked in both places

So, actually it does not matter where to release the request from performance perspective. I prefer to clean up resources where they are allocated if it's possible.

Reply:
ucp_request_release() is api function so it's not so good to call api function from internal flows (better use helper function), for example the api function can take a lock, this could cause lock ordering deadlock